chain/ethereum/src/block_stream.rs

```diff
@@ -383,20 +395,28 @@ where
                             let to_limit =
                                 cmp::min(head_ptr.number - reorg_threshold, next_start_block - 1);
 
-                            // The range should not be too small, due to the overhead of finding
-                            // triggers for each range, neither too large, so that progress is
-                            // commited frequently and memory usage is under control.
-                            let to = {
-                                let speedup = if from < *ETHEREUM_FAST_SCAN_END {
-                                    FAST_SCAN_SPEEDUP
-                                } else {
-                                    1
-                                };
-                                cmp::min(from + speedup * *ETHEREUM_BLOCK_RANGE_SIZE - 1, to_limit)
+                            // Calculate the range size according to the target number of triggers,
+                            // respecting the global maximum and also not increasing too
+                            // drastically from the previous block range size.
+                            let max_range_size =
+                                MAX_BLOCK_RANGE_SIZE.min(ctx.previous_block_range_size * 10);
+                            let range_size = if ctx.previous_triggers_per_block == 0.0 {
+                                max_range_size
+                            } else {
+                                (*TARGET_TRIGGERS_PER_BLOCK_RANGE as f64
+                                    / ctx.previous_triggers_per_block)
+                                    .max(1.0)
+                                    .min(max_range_size as f64)
+                                    as u64
```

Comment:
My brain is really bad at doing math unless it's written down. It would help if we had 2-3 examples in a comment here. Like what range do we get if there were a lot of triggers  previously, what if there were very few triggers and what if there was the ideal size?

```rust
// Examples (target triggers: 1000, max range size 10000, previous block range size 500):
// New max range size              -> min(10000, 5000 * 10) = 10000
// 1 triggers/block previously     -> min(max(1000 / 1, 1.0), 10000) = min(1000, 10000) = 1000
// 1000 triggers/block previously  -> min(max(1000 / 1000, 1.0), 10000) = min(1.0, 10000) = 1
// 10000 triggers/block previously -> min(max(1000 / 10000, 1.0), 10000) = min(1.0, 10000) = 1

Is there a mistake in my math or is the formula not ideal? In the second example, we're at the target triggers/block rate, so shouldn't the new block range size be exactly the same as before?

Reply:
I've added some examples in a comment.

If there are 1000 triggers per block in 1000 blocks that's 1 million triggers total, which is way above the target that's why the range decreases to 1 in your example.

Reply:
Oh, yeah, that part I got completely wrong. 1000 triggers as the target means 1000 triggers per _block_ range_, not per block. 🤦‍♂